feat: Add DVC support for large model files and configurable request timeout

teri0411 opened the pull request:
## Summary

  This PR adds DVC (Data Version Control) integration for handling large model weight files and introduces configurable request timeout for BentoML services.

  ## Features

  ### DVC Integration for Large Files
  - Automatically detect and upload large weight files (.pt, .bin, .pkl, etc.) to DVC remote storage
  - Files exceeding size threshold (default: 100MB) are managed via DVC instead of MLflow
  - Store DVC file hashes in MLflow tags for retrieval during serving
  - Restore DVC files automatically when loading models

  ### Configurable Request Timeout
  - Add `REQUEST_TIMEOUT` environment variable to configure BentoML service timeout
  - Default: 60 seconds

  ### S3-Compatible Storage Support
  - Add `AWS_ENDPOINT_URL` support for MinIO and other S3-compatible storage

  ### Performance Improvements
  - Multipart download for large S3 files (10 concurrent threads, 8MB chunks)

  ## Environment Variables

  | Variable | Description | Default |
  |----------|-------------|---------|
  | `DVC_ENABLED` | Enable DVC for large files | `false` |
  | `DVC_REMOTE_URL` | DVC remote storage URL | - |
  | `DVC_SIZE_THRESHOLD` | Min file size for DVC (bytes) | `104857600` (100MB) |
  | `REQUEST_TIMEOUT` | Request timeout in seconds | `60` |
  | `AWS_ENDPOINT_URL` | Custom S3 endpoint (optional) | - |

  ## Bug Fixes
  - Fix symlink creation error when target doesn't exist (DVC files restored later)
  - Fix symlink creation error when parent directory doesn't exist


  ---

teri0411 committed:
feat: set timeout env

teri0411 committed:
feat: add DVC support for large weight files and configurable request timeout

- Add Storage module for DVC operations (push/pull with retry logic)
- Integrate DVC with model registration (auto-detect large weight files)
- Add DVC file restoration during model serving
- Add REQUEST_TIMEOUT environment variable for BentoML service
- Add path traversal prevention for security

teri0411 committed:
feat: add DVC auto-initialization and environment validation

- Add automatic DVC init (--no-scm) when push_to_dvc is called
- Add automatic DVC remote configuration from DVC_REMOTE_URL
- Add AWS credentials validation for S3 remotes
- Add dvc, boto3, PyYAML dependencies to pyproject.toml
- Improve error messages for missing environment variables

Required env vars for S3:
- DVC_REMOTE_URL: s3://bucket/path
- AWS_ACCESS_KEY_ID
- AWS_SECRET_ACCESS_KEY

teri0411 committed:
feat: add AWS_ENDPOINT_URL support and refactor DVC operations

- Add AWS_ENDPOINT_URL env var for MinIO/S3-compatible storage
- Refactor push_to_dvc() to use dvc.repo.Repo Python API
- Add endpoint_url support to _download_from_s3()
- Refactor _ensure_dvc_remote_configured() with helper functions
- Export new AWS environment variable constants

teri0411 committed:
fix: skip symlink for non-existent targets in prepare_artifacts

Skip symlink creation when target file doesn't exist, as DVC files
are restored after model loading.

teri0411 committed:
test: add REQUEST_TIMEOUT validation tests

Add tests to verify REQUEST_TIMEOUT environment variable is properly
applied to BentoML service traffic timeout configuration.

teri0411 committed:
fix: restore makedirs for parent directory creation in prepare_artifacts

Parent directory must be created before symlink when artifact path
contains subdirectories (e.g., src/model.py).

teri0411 committed:
perf: add multipart download for faster S3 transfers

Use boto3 TransferConfig with multipart download for large files:
- multipart_threshold: 8MB
- max_concurrency: 10 threads
- multipart_chunksize: 8MB

teri0411 committed:
refactor: simplify DVC functions with dispatcher pattern and helper extraction

- Extract _parse_model_uri() and _get_model_version() from _get_dvc_info_from_model_uri()
- Add _needs_dvc_restoration() to simplify complex condition chains
- Add _get_downloader() dispatcher pattern to _download_from_remote()
- Deduplicate AWS credential validation with error_class parameter
- Move exception classes to top of Storage.py
- Update tests to use correct mock paths

teri0411 committed:
refactor: simplify DVC-related code and remove redundant comments

- Extract common iter_artifact_files generator to reduce duplication
- Simplify find_weight_files and _make_artifact_dir_dict using shared iterator
- Refactor _get_dvc_info_from_model_uri with helper function to reduce nesting
- Simplify load_nubison_mlflow_model by extracting helper functions
- Extract _parse_s3_url and _download_s3_with_cli from _download_from_s3
- Remove redundant inline comments throughout DVC-related code
- Reduce code by ~40% while maintaining same functionality

teri0411 committed:
fix: restore makedirs for parent directory creation in prepare_artifacts

Changed approach from pre-checking path.exists() to try-except pattern.
The previous implementation skipped symlink creation when target path
didn't exist, which broke artifact symlinks for valid files. Now creates
symlink first and catches FileNotFoundError for DVC files that will be
restored later.

teri0411 committed:
fix: preserve directory structure in artifacts and improve DVC upload performance

- Fix artifact path to preserve directory structure (e.g., "src/model.py" instead of "model.py")
  This fixes "No module named 'src'" error when using test_client with DVC enabled
- Add DVC_JOBS environment variable to configure parallel upload threads
- Change push_to_dvc to batch add all files at once for better performance

teri0411 committed:
feat: add parallel download and consolidate imports

- Add ThreadPoolExecutor for parallel DVC downloads using DVC_JOBS env var
- Handle FileExistsError silently in prepare_artifacts
- Move imports to top of file (dvc.repo.Repo, urllib.request)
- Default download workers: cpu_count() or 4

teri0411 committed:
feat: add DVC_FILE_EXTENSIONS env var for custom weight file extensions

Users can now add custom file extensions via DVC_FILE_EXTENSIONS env var.
Example: DVC_FILE_EXTENSIONS=".npy,.npz,.gguf"

teri0411 committed:
feat: create symlinks for DVC-restored files

Add _create_dvc_symlinks() to create symlinks in current working
directory for files restored from DVC. This enables import of
DVC-tracked Python modules during model loading.

teri0411 committed:
fix: isolate model cache by URI to prevent cross-version conflicts

Add _get_model_cache_key() to generate unique cache keys per model URI.
Cache files (path, lock) now include model URI hash, preventing stale
cache from affecting different model versions.

teri0411 committed:
refactor: remove redundant AWS credential validation and S3 CLI fallback

- Remove _validate_aws_credentials function that blocked IAM role usage
- Remove _download_s3_with_cli fallback (boto3 is a required dependency)
- Move boto3 imports to module level for cleaner code
- Let boto3 handle credential chain (env vars, credentials file, IAM role)

teri0411 committed:
refactor: simplify DVC pull using native dvc pull command

- Replace custom download logic with dvc pull -j for parallel downloads
- Remove boto3 direct dependency (already included in dvc[s3])
- Remove ChecksumMismatchError (DVC handles checksum verification internally)
- Remove retry_with_backoff, calculate_file_md5, verify_file_checksum
- Remove _download_from_remote, _download_from_s3, _parse_s3_url, _download_with_dvc
- Remove AWS credential environment variable exports from __init__.py
- Update tests to reflect simplified implementation

teri0411 committed:
fix: handle broken symlinks in prepare_artifacts

- Check path.exists() before path.isdir() to avoid errors on broken symlinks
- Remove FileNotFoundError from exception catch to properly report errors

teri0411 committed:
fix: use existing .dvc files and run DVC in model_root

- Remove .dvc file generation in pull_from_dvc; expect existing files from MLflow artifacts
- Add error if .dvc file not found
- Run DVC commands from model_root directory using temporary_cwd
- Update tests to create .dvc files before pull

teri0411 committed:
fix: maintain folder structure in MLflow artifacts for DVC compatibility

teri0411 committed:
refactor: reorganize DVC tests and add comprehensive test coverage

- Move DVC-related tests from test_service.py to test_storage.py
- Add TestCustomWeightExtensions for DVC_FILE_EXTENSIONS env var
- Add TestPathTraversalPrevention for security validation
- Add TestIterArtifactFiles for artifact directory traversal
- Add TestServiceDvcFunctions for Service.py DVC functions
- Add TestDvcVersionCache for version-based cache isolation
- Add subdirectory path test for pull_from_dvc

josephnw commented on nubison_model/Service.py:
```diff
@@ -162,18 +333,21 @@ def build_inference_service(
-        mlflow_tracking_uri or getenv(ENV_VAR_MLFLOW_TRACKING_URI) or DEAFULT_MLFLOW_URI
+        mlflow_tracking_uri
+        or getenv(ENV_VAR_MLFLOW_TRACKING_URI)
+        or DEAFULT_MLFLOW_URI
```
If there is no conflict with other variables, it may be better to fix DEAFULT_MLFLOW_URI variable to DEFAULT_MLFLOW_URI.

teri0411 replied:
Good catch! Fixed the typo — no conflicts with other variables.

opdaehyun commented on nubison_model/Storage.py:
```diff
@@ -0,0 +1,385 @@
+    return DEFAULT_SIZE_THRESHOLD
+
+
+def get_weight_extensions() -> set:
```
If type specification is possible, it would be better to write it in detail in the form of def get_weight_extensions() -> Set[str]:

teri0411 replied:
The difference between -> set and -> set[str] only matters for static type checkers (mypy). Runtime behavior is identical. This is an optional improvement, not required.

opdaehyun commented on nubison_model/Storage.py:
```diff
@@ -0,0 +1,385 @@
+            f"DVC: Pulling {len(dvc_file_paths)} file(s) with {jobs} workers..."
+        )
+
+    result = subprocess.run(
```
Add and push use the Repo API, but I'm curious why only pull uses subprocess.
I'd appreciate any comments on this.

teri0411 replied:
The key reason is the .dvc file targeting mechanism. We need to pull specific files based on .dvc files downloaded from MLflow:

  dvc pull model.pt.dvc weights.bin.dvc

  This CLI pattern directly accepts .dvc file paths as targets, which fits our use case perfectly.

teri0411 commented on nubison_model/Model.py:
```diff
@@ -176,20 +197,49 @@ def _make_conda_env() -> dict:
-    artifact_dirs_from_param_or_env = (
+def _make_artifact_dir_dict(
+    artifact_dirs: Optional[str],
+    exclude_weight_files: bool = False,
```
Large weight files are excluded from MLflow to:

  1. Prevent download timeouts — MLflow's HTTP download isn't optimized for large files
  2. Let each tool focus on its strength — MLflow for metadata/versioning, DVC for large binaries (with retry, caching, parallel downloads)
  3. Avoid duplicate storage

  The threshold is configurable (DVC_SIZE_THRESHOLD) so users can adjust based on their network conditions and project requirements.

teri0411 committed:
fix: include model_uri in DVC error logs for debugging

- Add model_uri parameter to _restore_dvc_files()
- Include model_uri in error messages to identify which model caused failures
- Update import to use corrected DEFAULT_MLFLOW_URI

teri0411 commented on nubison_model/Service.py:
```diff
@@ -76,68 +97,218 @@ def _extract_and_cache_model_path(mlflow_model, path_file):
-        _, nubison_model = _load_model_with_nubison_wrapper(
-            mlflow_tracking_uri, mlflow_model_uri
-        )
+        logger.warning("Lock timeout, falling back to direct load")
```
Good suggestion! Added model_uri to error logs for easier debugging. (44b28bc)

teri0411 committed:
fix: correct DEAFULT_MLFLOW_URI typo to DEFAULT_MLFLOW_URI

Fixes the variable name typo in Model.py definition.

opdaehyun reviewed (approved):
LGTM.

teri0411 commented:
@claude
  ---
  1. Security (Command Injection, Credential Exposure)

  No action needed. Environment variables are from trusted sources, shell=False prevents injection, and AWS/DVC never output credentials to stderr.

  ---
  2. Error Handling (Broad Exception, YAML Returns None)

  Intentional. We use raise DVCPushError(...) from e to preserve stack trace. YAML None case is already handled in push_to_dvc() which raises DVCPushError.

  ---
  3. Performance (getsize, Directory Traversal, FileLock 300s)

  No concerns. stat() runs in microseconds, register() is one-time operation, and 300s timeout is a safety margin for rare edge cases.

  ---
  4. Code Quality (Duplicate needs_dvc, Unused validate_safe_path)

  Both are incorrect:
  - needs_dvc double-check is intentional double-checked locking pattern for multi-worker race condition prevention
  - validate_safe_path() is used at Storage.py:343

  ---
  5. Test Coverage

  Tests exist: test_storage.py has 69 unit tests (617 lines) covering DVC push/pull, path traversal, multi-worker scenarios.

  ---
  6. Minor Items

  - Typo fix: Already done ✅
  - Magic numbers, naming conventions: Intentional design choices

  ---
  Summary: No changes required.
